### PR TITLE
[REEF-782] Allow REEF .NET Client to setup security token for Hadoop

### PR DESCRIPTION
**Pull request by anupam128**
This addressed the issue by
 * Defining file names for token id and password handshake between .NET
 * client and java client
 * Defining new commandline parameters for token kind and token service
 * and passing it to Java launcher
 * Setting the tokens into user credentials when user expresses such
 * intent by passing the token kind

JIRA:
 [REEF-782](https://issues.apache.org/jira/browse/REEF-782)

### CODE REVIEW

**Comment by jwang98052** [comment]
I will look at it

File: lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryPrefixParameter.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Client.YARN
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryPrefixParameter.cs]: Similar to Local, should we create a folder Parameter under YARN and put those classes there?

File: lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenKindParameter.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    [NamedParameter("Security token kind.", defaultValue:"NULL")]
+    public sealed class SecurityTokenKindParameter : Name<string>
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenKindParameter.cs]: This class is not in the csproj file

File: lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenParameters.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    [NamedParameter("Security token kind.", defaultValue:"NULL")]
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenParameters.cs]: Duplicated with the one in SecurityTokenKindParameter class. You can remove the one in this class as per class one file is good.

File: lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs [hunk 4]
```diff
@@ -89,13 +99,19 @@ private void Launch(IJobSubmission jobSubmission, string driverFolderPath)
...
+                JavaClassName,
+                driverFolderPath, // arg: 0
+                jobSubmission.JobIdentifier, // arg: 1
+                jobSubmission.DriverMemory.ToString(), // arg: 2
+                javaParams.TcpPortRangeStart.ToString(), // arg: 3
+                javaParams.TcpPortRangeCount.ToString(), // arg: 4
+                javaParams.TcpPortRangeTryCount.ToString(), // arg: 5
+                javaParams.MaxApplicationSubmissions.ToString(), // arg: 6
+                javaParams.DriverRestartEvaluatorRecoverySeconds.ToString(), // arg: 7
+                _securityTokenKind, // arg: 8
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs]: Similar as others, you might want to wrap those three into ClrClient2JavaClientCuratedParameters.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java [hunk 5]
```diff
@@ -281,6 +299,15 @@ private void writeDriverHttpEndPoint(final File driverFolder,
   public static void main(final String[] args) throws InjectionException, IOException, YarnException {
     final YarnSubmissionFromCS yarnSubmission = YarnSubmissionFromCS.fromCommandLine(args);
     LOG.log(Level.INFO, "YARN job submission received from C#: {0}", yarnSubmission);
+    if (!yarnSubmission.getTokenKind().equalsIgnoreCase("NULL")) {
```
Comment (jwang98052) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Move it to Launch would make main() cleaner.

**Comment by jwang98052** [comment]
I completed my pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java (hunk 5 above)
Comment (anupam128) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Cannot do that. Please see the comment added in the next lines.

File: lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs (hunk 4 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs]: Discussed offline with Julia.

Putting this in javaParams is a good idea but that makes configuring the parameters and plumbing much more cumbersome. Further, these parameters will go away when [REEF-70](https://issues.apache.org/jira/browse/REEF-70) is completed.

File: lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenKindParameter.cs (hunk 2 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenKindParameter.cs]: Yes. Will delete this file.

**Comment by anupam128** [comment]
@jwang98052 Thanks for taking a look. I have responded to comments and made changes.

**Pull request closed without merging**